File: funcscript/parser/syntax/FuncScriptParser.GetPrefixOperator.cs
using FuncScript.Block;
using FuncScript.Funcs.Math;
using FuncScript.Model;
using Microsoft.VisualBasic;

namespace FuncScript.Core
{
    public partial class FuncScriptParser
    {
        static ExpressionBlockResult GetPrefixOperator(ParseContext context, int index)
        {
            int i = 0;
            (var oper, var opNode, i) = GetOperator(context, s_prefixOp.Select(x => x[0]).ToArray(), index);

            if (i == index)
            {
                return new ExpressionBlockResult(null, null, index);
            }

            var symbol = s_prefixOp.First(x => x[0] == oper)[1];
            var func = context.ReferenceProvider.Get(symbol);
            if (func == null)
            {
                context.SyntaxErrors.Add(new SyntaxErrorData(index, i - index, $"Prefix operator {oper} not defined"));
                return new ExpressionBlockResult(null, null, index);
            }

            i = SkipSpace(context, i).NextIndex;

            var operandRes = GetCallAndMemberAccess(context, i);
            if (operandRes.NextIndex == i)
            {
                context.SyntaxErrors.Add(new SyntaxErrorData(i, 0, $"Operant for {oper} expected"));
                return new ExpressionBlockResult(null, null, index);
            }

            i = SkipSpace(context, operandRes.NextIndex).NextIndex;

            var expBlock = new FunctionCallExpression
            {
                Function = new LiteralBlock(func),
                Parameters = new[] { operandRes.Block },
                CodePos = index,
                CodeLength = i - index,
            };

            var parseNode = new ParseNode(ParseNodeType.PrefixOperatorExpression, index, i - index,
            new[] { opNode, operandRes.ParseNode });

            return new ExpressionBlockResult(expBlock, parseNode, i);
        }
    }
}

File: ts/core/src/parser/syntax/FuncScriptParser.GetPrefixOperator.ts
import { ParseContext, ParseNode, ParseNodeType, s_prefixOp, SyntaxErrorData } from "../FuncScriptParser.Main";
import { GetCallAndMemberAccess } from "./FuncScriptParser.GetCallAndMemberAccess";
import { GetLiteralMatch } from "./FuncScriptParser.GetLiteralMatch";
import { GetOperator } from "./FuncScriptParser.GetOperator";
import { SkipSpace } from "./FuncScriptParser.SkipSpace";

export function GetPrefixOperator(context: ParseContext, index: number) {
    let i = 0;
    let oper: string | null = null;
    let opNode: ParseNode | null = null;

    const matchResult = GetOperator(context,  s_prefixOp.map(x=>x[0]),index);
    i=matchResult.NextIndex
    oper=matchResult.MatchedOp
    if (i === index) {
        return { ParseNode: null, NextIndex: index };
    }

    if (!oper) {
        context.SyntaxErrors.push(new SyntaxErrorData(index, i - index, `Prefix operator ${oper} not defined`));
        return { ParseNode: null, NextIndex: index };
    }
    i = SkipSpace(context, i).NextIndex;

    const operandRes = GetCallAndMemberAccess(context, i);
    if (operandRes.NextIndex === i) {
        context.SyntaxErrors.push(new SyntaxErrorData(i, 0, `Operand for ${oper} expected`));
        return { ParseNode: null, NextIndex: index };
    }

    i = SkipSpace(context, operandRes.NextIndex).NextIndex;

    const parseNode = new ParseNode(
        ParseNodeType.PrefixOperatorExpression,
        index,
        i - index,
        [matchResult.ParseNode!, operandRes.ParseNode!]
    );

    return { ParseNode: parseNode, NextIndex: i };
}